Create reusable action derived from turbo-ignore

taesungh opened the pull request:
Originally desired in HackAtUCI/zothacks-site#209 and HackAtUCI/irvinehacks-site#439.

### Changes
- Create new composite action improving upon [Turborepo's turbo-ignore](https://turbo.build/repo/docs/reference/turbo-ignore)
- Compare hashes at the task level rather than only filtering packages
- Provide a step output of a status flag, similar to the exit codes

### Testing
1. Tested with personal fork of [HackAtUCI/irvinehacks-site](https://github.com/HackAtUCI/irvinehacks-site) using a workflow with steps similar to the sample usage
2. Observed deployment could be skipped for a non-functional modification such as updating a README file
3. Observed deployment could not be skipped for a functional modification such as a relevant code change

taesungh committed:
Create reusable action derived from turbo-ignore

- Create new composite action improving upon Turborepo's turbo-ignore
- Compare hashes at the task level rather than only filtering packages
- Provide a step output of a status flag, similar to the exit codes

samderanova commented on super-turbo-ignore.js:
```diff
@@ -0,0 +1,185 @@
+ * @returns {boolean} whether or not the deployment can be skipped
+ */
+function superTurboIgnore(task, workspace) {
+	info(`Using Turborepo to determine if this project is affected by the commit...\n`);
```
Does this need to use backticks or do regular quotes work?

taesungh replied:
Was copied from source, but I suppose not.

samderanova commented on super-turbo-ignore.js:
```diff
@@ -0,0 +1,185 @@
+
+	// check for TURBO_FORCE and bail early if it's set
+	if (process.env.TURBO_FORCE === "true") {
+		info("`TURBO_FORCE` detected");
```
Matter of preference but might be clearer to say `TURBO_FORCE environment variable detected`

taesungh replied:
Was copied from source.

samderanova commented on super-turbo-ignore.js:
```diff
@@ -0,0 +1,185 @@
+// Derived from vercel/turborepo:packages/turbo-ignore/src/ignore.ts@29aeaf0
```
Is TS supported by actions? If so, any reason for using JS over TS?

taesungh replied:
Was for sake of simplicity since running TypeScript directly would mean going through `npx ts-node` when node is already available. I suppose a possible alternative is using [Deno](https://deno.com) or [Bun](https://bun.sh) which support running TypeScript directly. Otherwise, would need to configure this project to run `tsc` and release the artifact.

taesungh replied:
Switching to bun/bunx since it appears to be faster (around 6 seconds to 4 seconds), not counting installation time.

taesungh committed:
Switch from using Node and npx to bun and bunx

- Bun can directly execute TypeScript code instead of needing ts-node
  - Overall execution also seems slightly faster
- Add `package.json` for dev-dependency on `@types/node`
- Move types directly into script and fix type errors
- Remove `npm-cache-dependency-path` parameter in action inputs

samderanova commented on action.yml:
```diff
@@ -0,0 +1,44 @@
+outputs:
+  status:
+    description: "Status is 0 if changes can be ignored, 1 if changes should be deployed."
+    value: ${{ steps.run_script.outputs.status}}
```
Nitpick: extra space after `status`?

samderanova reviewed (changes requested):
Looks good to me other than the one small nitpick!

Were you able to test whether this works or not or do you plan on testing it in the IrvineHacks site repo?

taesungh committed:
Check formatting

taesungh commented:
Yes, tested with a personal fork of the IrvineHacks site (and a fork of this action's repository) as noted in the PR description.
